### PR TITLE
WIP: apps: Switch to using OSSL_STORE for loading keys, certs, ...

### LINKED ISSUE #1963: Adapt the openssl app to use STORE_load
When the [STORE](https://github.com/openssl/openssl/projects/1) library is in place, it's time to adapt the openssl app to use it

### PR DESCRIPTION
**Pull request by levitte**
This makes all the format specifiers largely irrelevant, except for
the ENGINE format when a key is given as a key ID only.  However, with
the engine loader, it's possible to change this:

    -engine {engineid} -key {keyid} -keyformat ENGINE

To this:

    -key engine://{engineid}/{keyid}

We retain the ENGINE format for the sake of compatibility.  All other
input formats are simply ignored.  Note that output formats is a
different story.

Fixes #1963

### CODE REVIEW

**Comment by petrovr** [comment]
Richard Levitte wrote:
> This makes all the format specifiers largely irrelevant, except for
> the ENGINE format when a key is given as a key ID only.  However, with
> the engine loader, it's possible to change this:
>
>      -engine {engineid} -key {keyid} -keyformat ENGINE
>
> To this:
>
>      -key engine://{engineid}/{keyid}
[SNIP]
Proposed modification is not store specific.

Let me quote in brief from PKIX-SSH manual 
("http://securebox.termoneplus.com/man5/ssh_config.5.html") for option 
IdentityFile.

(1) If file name starts with prefix ‘engine:’ instead from file identity 
load is redirected to “loadable cryptographic module” (engine).
Name format is engine:{engineid}:{keyid} using syntax above.

(2) Prefix “store:” could be used if cryptographic library supports 
ossl_store(7) functionality.
In such case name format is store:{uri}. Engine if applicable is encoded 
in URI-scheme.

Second case is more generic. It is store related as it allows use of 
URIs not supported by engine.
For instance based on store functionality PKIX-SSH could use directly as 
identify key and certificates stored in PKCS#12 file (side effect).
Without store user should unpack (extract) PKCS#12 file into text file 
(PEM format).


Otherwise there is no big difference for PKIX-SSH:
(1) ssh ... -i engine:e_nss:{keyid} vs
(2) ssh ... -i store:nss:{keyid}
Remark: in both cases engine require explicit configuration.


And one note more: {engineid}/{keyid} is not _in store URI format._  You 
know engine , you know keyid, but you don't know how to build store URI. 
It is engine specific and is more one direction process: URI -> 
{{engineid}, {keyid}, {extras...}}.

Regards,
Roumen Petrov

**Comment by levitte** [comment]
Ah, there's a précédent for a engine: scheme. Sure, I can use that instead.

Why someone has implemented a store: scheme is beyond me, though. Does that mean that if I want to express a file name in uri form, I would have to write `store:file:///home/levitte/foo.pem`? Why isn't `file:///home/levitte/foo.pem` enough? I cannot see why we should support such a contraption...

**Comment by petrovr** [comment]
Richard Levitte wrote:
> Ah, there's a précédent for a engine: scheme. Sure, I can use that instead.
> 
> Why someone has implemented a store: scheme is beyond me, though. Does that mean that if I want to express a file name in uri form, I would have to write `store:file:///home/levitte/foo.pem`?

So I use prefix:data, whre prefix is engine or store  just to avoid 
failback case into code.

There is no reason OpenSSL to use such format.


>  Why isn't `file:///home/levitte/foo.pem` enough? I cannot see why we should support such a contraption...
> 

Yep . Support of "Store URI" as input source should be enough . Perhaps 
if store fails with "invalid scheme" then code could try to open file 
(failback case).

Drawback of "Store URI" is that engine may not be loaded.
Work around is to keep "engine" argument.


Regards,
Roumen Petrov

**Comment by levitte** [comment]
> Perhaps if store fails with "invalid scheme" then code could try to open file (failback case).

The current file loader does this already, i.e. it defaults to `file:` if no scheme is present.

**Comment by levitte** [comment]
> Drawback of "Store URI" is that engine may not be loaded. Work around is to keep "engine" argument.

The proper workaround is to do what I do in this PR, i.e. define a `engine:` scheme loader.  See `apps/app_engine_loader.c`

**Comment by kaduk** [comment]
Please do not start using an unregistered URI scheme: https://www.iana.org/assignments/uri-schemes/uri-schemes.xhtml

**Comment by levitte** [comment]
Code updated to use `engine:{engineid}:{keyid}`

**Comment by levitte** [comment]
Apparently, I hadn't completed what I said.  Now fixed, and tests added.

**Comment by levitte** [comment]
Finally, this builds and tests fine. I will fix the docs, then it will be time for a review

**Comment by petrovr** [comment]
Richard Levitte wrote:
> Apparently, I hadn't completed what I said.  Now fixed, and tests added.
> 


Perhaps in next weeks :-( I could find time to continue discussion.

Regards,
Roumen

**Comment by dwmw2** [comment]
I'll echo the "please don't use an unregistered URI scheme" comment with its converse: please *do* support PKCS#11 URIs directly, without them having to be prefixed or mangled. The `{engine}:…` form mentioned above would have worked for that, given that the name of the engine in question would be `pkcs11`.

If you end up with a form that doesn't Just Work for RFC7512 PKCS#11 URIs, please special-case them to use the pkcs11 engine. Thanks.

**Comment by levitte** [comment]
If you create a proper OSSL_STORE loader that registers the scheme "pkcs11:", then you can use that scheme directly.

The "engine:" 'scheme' is an interim construct to allow access via the older ENGINE calls to access keys; `ENGINE_load_private_key` and `ENGINE_load_public_key`.  However, if the engine offers an OSSL_STORE loader, *there is no reason whatsoever to use the "engine:" 'scheme'*

**Comment by dwmw2** [comment]
Perfect. Thank you.

**Comment by mattcaswell** [comment]
Needs a rebase

**Comment by mattcaswell** [comment]
> But seeing that PKIX-SSH uses the same scheme, and uses a different syntax to boot, that might be qualification enough to make a provisional registration after all. (I don't see it ever changing status to something more provisionalpermanent, btw)

Was the scheme ever registered as suggested here?

File: crypto/store/loader_file.c [hunk 1]
```diff
@@ -163,7 +166,7 @@ typedef OSSL_STORE_INFO *(*file_try_decode_fn)(const char *pem_name,
                                                size_t len, void **handler_ctx,
                                                int *matchcount,
                                                const UI_METHOD *ui_method,
-                                               void *ui_data);
+                                               void *ui_data, const char *uri);
```
Comment (mattcaswell) [inline crypto/store/loader_file.c]: The comment above here needs updating

**Comment by levitte** [comment]
> Was the scheme ever registered as suggested here?

No, but can certainly be done.  I have the impression that it's not very hard...

File: doc/man1/openssl.pod [hunk 2]
```diff
@@ -491,6 +491,35 @@ Provides a terse summary of all options.
...
+Keys that are offered by an engines are an exception, they can still
+be given like this:
+
+ -engine eng -key something -keyform ENGINE
+
+The exact same specification can also be given like this:
+
+ -key engine:eng:something
+
+This is a temporary measure until engines have learned to provide an
```
Comment (mattcaswell) [inline doc/man1/openssl.pod]: I somehow suspect this won't be temporary and this wording will still be here in 10 years time. Perhaps reword?

File: util/libcrypto.num [hunk 3]
```diff
@@ -4799,3 +4799,5 @@ EC_GROUP_check_named_curve              4746	3_0_0	EXIST::FUNCTION:EC
 EVP_CIPHER_upref                        4747	3_0_0	EXIST::FUNCTION:
 EVP_CIPHER_fetch                        4748	3_0_0	EXIST::FUNCTION:
 EVP_CIPHER_mode                         4749	3_0_0	EXIST::FUNCTION:
+OSSL_STORE_attach                       4750	3_0_0	EXIST::FUNCTION:
+OSSL_STORE_LOADER_set_attach            4751	3_0_0	EXIST::FUNCTION:
```
Comment (mattcaswell) [inline util/libcrypto.num]: These functions don't seem to be documented.

File: test/recipes/90-test_store.t [hunk 4]
```diff
@@ -152,6 +164,25 @@ indir "store_$$" => sub {
...
+		ok(run(app(["openssl", "rsa", "-text", "-noout", @pubin,
+			    "-engine", "ossltest", "-inform", "engine",
+			    "-in", "ot:".$file])));
+		ok(run(app(["openssl", "rsa", "-text", "-noout", @pubin,
+			    "-engine", "ossltest", "-inform", "engine",
+			    "-in", "ot:".to_abs_file($file)])));
+		ok(run(app(["openssl", "rsa", "-text", "-noout", @pubin,
+			    "-in", "engine:ossltest:ot:$file"])));
+		ok(run(app(["openssl", "rsa", "-text", "-noout", @pubin,
+			    "-in", "engine:ossltest:ot:".to_abs_file($file)])));
```
Comment (mattcaswell) [inline test/recipes/90-test_store.t]: Given the extent of the option changes in the apps this seems very light testing

**Review by mattcaswell — CHANGES_REQUESTED** [review]
I'd also suggest a CHANGES entry might be in order for the change in behaviour of "-inform"

File: apps/app_engine_loader.c [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+
+    if ((loader = OSSL_STORE_LOADER_new(NULL, "engine")) == NULL
+        || !OSSL_STORE_LOADER_set_open(loader, engine_open)
+        || !OSSL_STORE_LOADER_set_ctrl(loader, engine_ctrl)
+        || !OSSL_STORE_LOADER_set_load(loader, engine_load)
+        || !OSSL_STORE_LOADER_set_eof(loader, engine_eof)
+        || !OSSL_STORE_LOADER_set_error(loader, engine_error)
+        || !OSSL_STORE_LOADER_set_close(loader, engine_close)
+        || !OSSL_STORE_register_loader(loader)) {
+        OSSL_STORE_LOADER_free(loader);
```
[the remark this answers is not in the document]
Comment (levitte) [reply apps/app_engine_loader.c]: Considering we have declared ENGINE legacy, I dunno how keen I'm on that.  But, we could re-evalutate that when I've finished #9389

**Comment by levitte** [comment]
> Needs a rebase

Getting at it now

**Comment by levitte** [comment]
The more I looked at the branch, the more I realised it's not quite ready for final review.  So I put it back in WIP.

Comments are still welcome!

**Comment by t-j-h** [comment]
This is still a WIP but placing a hold so it doesn't end up in OpenSSL 3.0 without a discussion and decision on the points already raised in the feedback.

**Comment by DDvO** [comment]
This is related to #12647 and likely further more recent PRs..

**Comment by t8m** [comment]
Isn't this change already implemented in master/3.0?

**Comment by levitte** [comment]
> Isn't this change already implemented in master/3.0?

Almost, there were nuances that are not caught by @DDvO's changes.
I've left this unchanged for a while for the sake of priority, but intended to rebase this when I've got a chunk of time, to see what there is left.

**Comment by mattcaswell** [comment]
OMC: There is no decision to be made here at this stage.

**Pull request closed without merging**